Supports viewing class member variables in lambda when using the vs debugger

GkvJwa opened the pull request:
Use "__this" in DataMemberRecord, make vs debugger can be parsed normally

Fixes #71562

GkvJwa committed:
Supports viewing class member in lambda when using the vs debugger

Use "__this" in the DataMemberRecord when generating pdb, so that the vs debugger can parse the  member vars in lambda normally

dwblaikie commented on llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp:
```diff
@@ -2605,6 +2605,9 @@ CodeViewDebug::lowerRecordFieldList(const DICompositeType *Ty) {
       MemberBaseType = TypeTable.writeLeafType(BFR);
     }
     uint64_t MemberOffsetInBytes = MemberOffsetInBits / 8;
+    if (Ty->getName().find("<lambda") != std::string::npos) {
+      MemberName = "__this";
+    }
```
This should probably be done by checking whether the `Ty` is a lambda, rather than searching through its name. Oh, this is down in LLVM, so we wouldn't have any info about this.

So maybe this should be up in clang? Hmm, even compiling for a windows target, I don't see a member for the `this` pointer in the LLVM IR metadata - we only get members for the captures, by the looks of it... 

Maybe this MemberInfo stuff is PDB specific/maybe it synthesizes a this pointer from somewhere else... 

I'll leave this to folks more familiar with the PDB side of things.

GkvJwa replied:
Yeah, I debugged and saw that when parsing the AST the pdb information is generated. Add member information is also here

looking at the lambda function, its naming rules are as follows
https://github.com/llvm/llvm-project/blob/edd690b02e16e991393bf7f67631196942369aed/clang/lib/AST/MicrosoftMangle.cpp#L250-L255

Therefore, if want to modify the place where the information is generated, I don’t know if there is any other way to know that it is a lambda

ZequanWu replied:
This is inside a for loop iterating the class members. Doesn't this changes all class members name belonging to a lambda to "__this"?

ZequanWu replied:
Just realized that this fix should be done in Clang:
https://github.com/llvm/llvm-project/blob/0e42df4031e8b2fec357e07ca5ca3b81adf0b5ad/clang/lib/CodeGen/CGDebugInfo.cpp#L1660-L1662

When emitting codeview, use `__this` for the captured `this`.
```
      StringRef ThisName = CGM.getCodeGenOpts().EmitCodeView ? "__this": "this";
      llvm::DIType *fieldType = createFieldType(
          ThisName, type, f->getLocation(), f->getAccess(),
          layout.getFieldOffset(fieldno), VUnit, RecordTy, CXXDecl);
```

Then, the test can be a C++ file put inside `clang/test/CodeGenCXX`.

GkvJwa replied:
Good, I tried patching it here and it also works, and the pdb compiled by msvc also has operators:
```
 0x465D | LF_FIELDLIST [size = 160, hash = 0x26758]
          - LF_ONEMETHOD [name = `operator()`]
            type = 0x4653, vftable offset = -1, attrs = public
          - LF_METHOD [name = `<lambda_1>`, # overloads = 2, overload list = 0x4658]
          - LF_MEMBER [name = `__this`, Type = 0x464E, offset = 0, attrs = private]
          - LF_MEMBER [name = `aa`, Type = 0x0074 (int), offset = 8, attrs = private]
          - LF_MEMBER [name = `bb`, Type = 0x0074 (int), offset = 12, attrs = private]
          - LF_MEMBER [name = `cc`, Type = 0x0074 (int), offset = 16, attrs = private]
          - LF_ONEMETHOD [name = `operator=`]
            type = 0x465C, vftable offset = -1, attrs = public
          - LF_ONEMETHOD [name = `__autoclassinit2`]
```
Not sure what it means

rnk commented:
This also needs a test. I booted up my Windows machine to try to better understand what needs to be done, but I had to prioritize other work in the meantime and haven't gotten around to it.

GkvJwa commented:
> This also needs a test. I booted up my Windows machine to try to better understand what needs to be done, but I had to prioritize other work in the meantime and haven't gotten around to it.

You can create a new project through vs, use msvc(v143) and llvm to compile the code in the issue , You should be able to reproduce this
`Properties -> General -> Platform toolset -> (v143/llvm)`

And how to add a test, which affects whether the debugger can recognize

ZequanWu reviewed (commented):
> And how to add a test, which affects whether the debugger can recognize

In terms of test, I think you can add a test to check if the generated member name is "__this" using `llvm-pdbutil` and run check-llvm.

GkvJwa commented:
> > And how to add a test, which affects whether the debugger can recognize
> 
> In terms of test, I think you can add a test to check if the generated member name is "__this" using `llvm-pdbutil` and run check-llvm.

I tried to use the following code for testing(add some member variables):
```
class foo {
 public:
  void fun() {
    int aa = 1;
    int bb = 2;
    auto f = [=] {
      printf("%d %d\n", a, b);
	  printf("%d %d\n", aa, bb);
    };
    f();
  }
  ....
};
```
The information about other member variables was also modified to __this, which can also be parsed by the vs debugger

But should only modify the implicit this, so I added a check

---
In addition, I tried to use static lamdba and found that vs did not seem to support it, so I skipped overriding the variables here
https://github.com/llvm/llvm-project/blob/edd690b02e16e991393bf7f67631196942369aed/llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp#L2575-L2585

ZequanWu commented on llvm/test/DebugInfo/COFF/lambda-this-info.ll:
```diff
@@ -0,0 +1,245 @@
+
+; CHECK:                       Types (.debug$T)
+; CHECK-NEXT: ============================================================
+; CHECK:0x1003 | LF_CLASS [size = 36] `Foo`
```
It's better to avoid hard coded indexes like `0x1003` etc and using [regex](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks). 

Consider using `[[FooIndex:0x[A-Z0-9]+]]`

GkvJwa replied:
I use regular expressions to identify index, but type is still hard-coded

The type will remain consistent with this:
https://github.com/llvm/llvm-project/blob/edd690b02e16e991393bf7f67631196942369aed/llvm/test/DebugInfo/COFF/udts-complete.ll#L20-L30

ZequanWu commented on clang/test/CodeGenCXX/lambda-this-info.cpp:
```diff
@@ -0,0 +1,48 @@
+// CHECK-NEXT: ============================================================
+// CHECK:[[FooIndex:0x[^ ]*]] | LF_CLASS [size = 36] `Foo`
+// CHECK:         unique name: `.?AVFoo@@`
+// CHECK:[[FooIndex:0x[^ ]*]] | LF_FIELDLIST [size = 52]
```
Nit: You are creating multiple `FooIndex` but not using it. The purpose of regex string substitution is to capture a pattern that will show up later. Different patterns should have different names. Here, it could be `[[FooFieldIndex:0x[^ ]*]]` In order to use the patterns later, use `0x[[FooFieldIndex]]`. Same applies to other indexes.

dwblaikie commented on clang/test/CodeGenCXX/lambda-this-info.cpp:
```diff
@@ -0,0 +1,48 @@
+// RUN: %clang_cl --target=x86_64-windows-msvc /c /Z7 -o %t.obj -- %s
```
Generally clang tests don't test end-to-end. If the code change only affects clang IR generation, the test should only test clang IR generation, not all the way down to PDB, etc.

GkvJwa replied:
I check the IR through the command line:
```
%clang_cc1 -triple x86_64-windows-msvc -emit-llvm --std=c++17 %s -o -
```
and confirmed that this patch will not be reflected in the IR. So, I think it would be better to be placed in ```llvm/test/DebugInfo/COFF```

ZequanWu replied:
I think you are missing `-gcodeview` in the command line to emit the debug info. The change should reflect on a metadata. `llvm/test/DebugInfo/COFF` is the place to test the phase when IR being lowered to object file.  `clang/test/CodeGenCXX` is for testing C++ code to IR.

GkvJwa replied:
got it, review again

dwblaikie reviewed (approved):
Looks plausible to me, thanks!

GkvJwa commented:
Help me merge it into the main branch, Thanks
